### PR TITLE
Implement GSP-134 and GSP-654

### PR DESCRIPTION
**Pull request by JinnyYi**

### CODE REVIEW

File: storage.go [hunk 1]
```diff
@@ -42,10 +42,24 @@ func (s *Storage) createDir(ctx context.Context, path string, opt pairStorageCre
 
 	rp := s.getAbsPath(path)
 
-	// Add `/` at the end of path to simulate directory.
+	// Add `/` at the end of `path` to simulate a directory.
 	// ref: https://developer.qiniu.com/kodo/kb/1705/how-to-create-the-folder-under-the-space
 	rp += "/"
 
+	// kodo `put` doesn't support `overwrite`, so we need to check and delete the dir if exists.
```
Comment (JinnyYi) [inline storage.go]: Maybe we should omit resource  exist error here and there's not need to delete and recreate the dir?

File: storage.go [hunk 2]
```diff
@@ -267,12 +287,26 @@ func (s *Storage) stat(ctx context.Context, path string, opt pairStorageStat) (o
...
+	if err == nil {
+		err = s.bucket.Delete(s.name, rp)
+	}
+	if err != nil && checkError(err, responseCodeResourceNotExist) {
+		err = nil
+	}
+
+	if err != nil {
+		return
+	}
```
Comment (xxchan) [inline storage.go]:
```suggestion
	if err == nil {
		err = s.bucket.Delete(s.name, rp)
		if err != nil {
			return
		}
	} else if !checkError(err, responseCodeResourceNotExist) {
		return
	}
```

Will this look clearer (and equivalent)?

File: storage.go [hunk 3]
```diff
@@ -42,10 +42,24 @@ func (s *Storage) createDir(ctx context.Context, path string, opt pairStorageCre
...
```
Comment (xxchan) [inline storage.go]: same as below

File: storage.go (hunk 1 above)
Comment (xxchan) [reply storage.go]: This seems reasonable (but not specified?). Sounds like `mkdir -p` (existing directory and contents are preserved)

File: storage.go (hunk 3 above)
Comment (xxchan) [reply storage.go]: (can remove if unnecessary

**Review by xxchan — APPROVED** [review]

**Pull request merged**